pkg/runner/runner.go

```diff
@@ -0,0 +1,199 @@
+}
+
+func (o *Runner) CheckUpkeep(ctx context.Context, mercuryEnabled bool, keys ...ocr2keepers.UpkeepKey) ([]ocr2keepers.UpkeepResult, error) {
+	r, err := o.parallelCheck(ctx, mercuryEnabled, keys)
+	if err != nil {
+		return nil, err
+	}
+
+	return r.Values(), nil
+}
+
+func (o *Runner) Start() error {
+	if !o.running.Load() {
+		go o.cacheCleaner.Run(o.cache)
+		o.running.Swap(true)
+	}
+
+	return nil
+}
+
+func (o *Runner) Close() error {
+	if o.running.Load() {
+		o.cacheCleaner.Stop()
+		o.workers.Stop()
+		o.running.Swap(false)
+	}
+
+	return nil
+}
+
+// parallelCheck should be satisfied by the Runner
+func (o *Runner) parallelCheck(ctx context.Context, mercuryEnabled bool, keys []ocr2keepers.UpkeepKey) (*Result, error) {
+	result := NewResult()
+
+	if len(keys) == 0 {
+		return result, nil
+	}
+
+	toRun := make([]ocr2keepers.UpkeepKey, 0, len(keys))
+	for _, key := range keys {
+
+		// if in cache, add to result
+		if res, ok := o.cache.Get(string(key)); ok {
+			result.Add(res)
+			continue
+		}
+
+		// else add to lookup job
+		toRun = append(toRun, key)
+	}
+
+	// no more to do
+	if len(toRun) == 0 {
+		return result, nil
+	}
+
+	// Create batches from the given keys.
+	// Max keyBatchSize items in the batch.
+	pkgutil.RunJobs(
+		ctx,
+		o.workers,
+		util.Unflatten(toRun, o.workerBatchLimit),
+		o.wrapWorkerFunc(mercuryEnabled),
+		o.wrapAggregate(result),
+	)
+
+	if result.Total() == 0 {
+		o.logger.Printf("no network calls were made for this sampling set")
+	} else {
+		o.logger.Printf("worker call success rate: %.2f; failure rate: %.2f; total calls %d", result.SuccessRate(), result.FailureRate(), result.Total())
+	}
+
+	// multiple network calls can result in an error while some can be successful
+	// in the case that all workers encounter an error, bubble this up as a hard
+	// failure of the process.
+	if result.Total() > 0 && result.Total() == result.Failures() && result.Err() != nil {
+		return nil, fmt.Errorf("%w: last error encounter by worker was '%s'", ErrTooManyErrors, result.Err())
+	}
+
+	return result, nil
+}
+
+func (o *Runner) wrapWorkerFunc(mercuryEnabled bool) func(context.Context, []ocr2keepers.UpkeepKey) ([]ocr2keepers.UpkeepResult, error) {
+	return func(ctx context.Context, keys []ocr2keepers.UpkeepKey) ([]ocr2keepers.UpkeepResult, error) {
+		start := time.Now()
+
+		// perform check and update cache with result
+		checkResults, err := o.registry.CheckUpkeep(ctx, mercuryEnabled, keys...)
+		if err != nil {
+			err = fmt.Errorf("%w: failed to check upkeep keys: %s", err, keys)
+		} else {
+			o.logger.Printf("check %d upkeeps took %dms to perform", len(keys), time.Since(start)/time.Millisecond)
+
+			for _, result := range checkResults {
+				ok, err := o.encoder.Eligible(result)
+				if err != nil {
+					o.logger.Printf("eligibility check error: %s", err)
+					continue
+				}
+
+				if ok {
+					key, _, _ := o.encoder.Detail(result)
+					o.logger.Printf("upkeep ready to perform for key %s", key)
+				}
```

Comment:
nit. can we also log the opposite (when an upkeep is not eligible)

We do this in main https://github.com/smartcontractkit/ocr2keepers/blob/main/internal/keepers/service.go#L318-L320